### PR TITLE
Comments: move selected comment state to redux

### PR DESCRIPTION
**Pull request by rodrigoi**
Depends on #18995.

Selecting comments, and the state of each individual comment, is a piece of shared data across different levels of the component tree, currently managed by `CommentList` as local state.
This PR moves that state to the `UI` redux tree, adding the actions to modify that state and the selectors to access it. `CommentList` still sets the status of the selected comments.

This PR keeps 1:1 parity with the private state implementation. There should be no regressions.

### State Tree Changes

The selected state for each comment is added under the `UI` redux tree, keyed by `siteId`. For example:

```js
{
  ui: {
      comments: {
          selected: {
            123456: {
              1701: { commentId: 1701, postId: 1 },
              1138: { commentId: 1138, postId: 1 }
            }
          }
      }
  }
}
```

### Actions

Three new actions are added. `COMMENTS_TOGGLE_SELECTED`, to add or remove a single comment from the state tree. `COMMENTS_CLEAR_SELECTED` to clear the state for the specified `siteId`, and `COMMENTS_SELECT_ALL` to replace the selected comments for the specified `siteId` with a provided list of comments.

### Selectors

`getSelectedComments` is the basic selector to access all selected comments for a `siteId`. `getSelectedCommentsCount` and `isCommentSelected` use that root selector to get the selected count for a `siteId`, or to access the state of a single comment.

### Next steps

- Add a bulk actions endpoint.
- Create the `data-layer` actions for that endpoint.
- Remove `CommentList` code that manages side effects and move it to the `data-layer`.

### References:

Exploratory testing on #18932, including `data-layer` handlers. Bulk actions endpoint (in progress) on D7771-code.

### CODE REVIEW

File: client/my-sites/comments/comment-navigation/index.jsx [hunk 1]
```diff
@@ -249,11 +249,16 @@ const mapStateToProps = ( state, { commentsPage, siteId } ) => {
 			};
 		}
 	} );
+	const selectedComments = getSelectedComments( state, siteId );
 
 	return {
+		isSelectedAll:
+			difference( commentsPage, map( selectedComments, ( { commentId } ) => commentId ) ).length ===
```
Comment (Copons) [inline client/my-sites/comments/comment-navigation/index.jsx]: I assume this is in case some or all `selectedComments` are not contained in the current page.
Is it something that can actually happen _right now_, or is it future-proofing in case we add multi-page selection?

Comment (rodrigoi) [reply client/my-sites/comments/comment-navigation/index.jsx]: You hit the jackpot. Having external state makes it really easy to have selections across pages. The original code just checked the `length` of the selected comments to be the same as the page size. This, in my opinion, is more robust. Performance wise, we can add memoization later if it turns out to be a problem.

File: client/my-sites/comments/comment-navigation/index.jsx [hunk 2]
```diff
@@ -263,8 +268,19 @@ const mapDispatchToProps = {
 			recordTracksEvent( 'calypso_comment_management_change_filter', { status } ),
 			bumpStat( 'calypso_comment_management', 'change_filter_to_' + status )
 		),
+	clearSelectedComments,
+	selectAllComments,
 };
 
-export default connect( mapStateToProps, mapDispatchToProps )(
+const mergeProps = ( stateProps, dispatchProps, ownProps ) => ( {
```
Comment (Copons) [inline client/my-sites/comments/comment-navigation/index.jsx]: TIL! 🌮

File: client/my-sites/comments/comment-navigation/index.jsx (hunk 1 above)
Comment (Copons) [reply client/my-sites/comments/comment-navigation/index.jsx]: 💰

**Comment by Copons** [comment]
Noticed that we are too relaxed in clearing the selected comments (it only happens on `changePage`, as far as I can see).

- Let's say we are in the All list, select all (1 pending, 19 approved).
- Change to Pending list: the only comment is still selected; the select all checkbox is ticked; the selected count shows 20.
- Change to Approved: 19 out of 20 comments are selected; the select all checkbox is unticked; the selected count still shows 20.
- Change to Spam: 0 comments are selected; the select all checkbox is unticked; the selected count still shows 20.
- With 0 spam comments selected, but with a selected count > 0, the bulk action buttons (based on the current list: since we're on Spam, we only have Approve and Delete) are enabled, and we can apply them on selected comments not included in the current list.

Imho, the selected comments should be emptied every time we change page or route.

File: client/state/ui/comments/selected/reducer.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+ * Internal dependencies
+ */
+import {
+	COMMENTS_CLEAR_SELECTED,
+	COMMENTS_SELECT_ALL,
+	COMMENTS_TOGGLE_SELECTED,
+} from 'state/action-types';
+import { keyedReducer } from 'state/utils';
+
+const mapComments = partialRight( map, ( { commentId, postId } ) => ( { commentId, postId } ) );
```
Comment (Copons) [inline client/state/ui/comments/selected/reducer.js]: Do we really need `partialRight` here?
Wouldn't it be more readable by simply using `map` below?

**Comment by Copons** [comment]
This looks awesome @rodrigoi!
My first review produced only nitpicks, and a UX question.
Haven't looked at the tests yet as I'm bad at them. 🙂

**Comment by kwight** [comment]
Are we certain we need this? I can see improving how `<CommentList>` handles its local state and other components, but I'm not sure of the value of moving this to the store... I personally wouldn't expect to see still-selected comments upon returning to a site list (in fact, I'd prefer it to be clean with nothing selected, so that I don't have to reset anything myself). It feels like a great candidate for leaving it in local state, rather than adding more unnecessarily to the store.

That said, it works beautifully :)

**Comment by Copons** [comment]
@kwight The value is that we need to know about selected comments in two different places: the navigation bar and the comment detail header.
What we do now is to manage everything from `CommentList` which is currently super bloated.

Though, I share your same concerns about clearing the selected list.
Maybe we could simply clear it every time we land on a `/comments` route?

**Comment by kwight** [comment]
> What we do now is to manage everything from CommentList which is currently super bloated, and could really use a lot of clean up.

I think that's my point; our real issue is that CommentsList needs refactoring and cleanup (having a HOC with local state that handles things for child components is still a valid way of working). We're moving bulk edit to the store as a fix for that issue, not because there's a justification for having it in the store (justification being, we need that data to persist between restarts or page navigation). Just my 2￠.

**Comment by kwight** [comment]
Oh, got myself into this situation, which is confusing:

screen shot 2017-10-19 at 11 19 26 am

**Comment by gwwar** [comment]
> having a HOC with local state that handles things for child components is still a valid way of working

Also, true. Playing around with this a bit more it may make more sense for it to live within component state, since as @kwight mentioned, we have no use cases yet for the data being persisted beyond the component unmount, or for other components to interact with this UI state.

File: client/my-sites/comments/comment-navigation/index.jsx (hunk 1 above)
Comment (gwwar) [reply client/my-sites/comments/comment-navigation/index.jsx]: @megsfulton do you see us having a need for a select all? Eg where we have say 299 unapproved comments and it selects all 299 of them, vs the first 20 on the page?

Comment (megsfulton) [reply client/my-sites/comments/comment-navigation/index.jsx]: @gwwar I'm not sure, but we can add it to our list of questions and keep an eye on the stats for bulk actions.

**Comment by Copons** [comment]
So, I've spent a bit of time wondering how to convert this into a HOC.
You can see the result [here](https://github.com/Automattic/wp-calypso/blob/f35d35c55cb25ceaabb31fca243c097687cc6720/client/my-sites/comments/comment-list/comments-manager.jsx), on a branch that I willingly won't convert into a PR.

The idea was: let's move the bulk stuff away from `CommentList` and into a HOC that then will be applied to both `CommentList` and `CommentNavigation`.
Though, bulk stuff depends on the list of comments, which also depends on the persistence logic.
So when I started moving methods away from `List` and into the HOC, I started a chain reaction, and eventually I had to move half the `List` methods.
Which it's a good thing by itself: the `List` is a bloated mess, and it could really use some _extreme makeover_.

Eventually, if you look at the HOC, you notice that it looks pretty much identical to `List`. I just moved the problem elsewhere.

But more importantly, I was missing the whole point of @rodrigoi's PR: having a single source of truth for both `List` and `Navigation`.
The HOC internal state scope would only be limited to either `List` or `Navigation` but not both.
Unless... we applied it to a component containing both `List` and `Navigation. Such component currently exists, and it's the `List` itself.
Again, moving the problem elsewhere.

All in all, I think we're back with two options: the current one, Rodrigo's full Redux.
If the reason for the HOC is to reduce `List`'s bloat, well, just imagine that most of its methods will be removed in the full Redux refactor of `CommentDetail`.

tl;dr after giving the HOC a try, I'm fully backing this PR.

File: client/my-sites/comments/comment-navigation/index.jsx (hunk 1 above)
Comment (rodrigoi) [reply client/my-sites/comments/comment-navigation/index.jsx]: we've discussed the possibility of the checkbox selecting the _current page_ + a message asking if the user want to select all, like gmail does. But indeed it's a question for power users. I think that for the rest of us, quick actions would be more useful.

**Comment by kwight** [comment]
@Copons Oh, I wasn't suggesting adding an HOC on top of what we have now (yes, definitely way more messy and complicated), but rather refactor and clean up CommentsList so that it continues to act as an HOC for the two things that need it, but in a much better and easier-to-reason-about way. My main point is that we should only be moving something into the store if data needs to persist between refreshes and page navigation, not to try and compensate for a component in need of refactoring.

Let me take a crack at demonstrating how I think we can do this 👍

**Comment by vindl** [comment]
> But more importantly, I was missing the whole point of @rodrigoi's PR: having a single source of truth for both List and Navigation. The HOC internal state scope would only be limited to either List or Navigation but not both.

That's what I was thinking too while reading through HOC proposal. We'd just end up with two wrapped components with separate component state.

> My main point is that we should only be moving something into the store if data needs to persist between refreshes and page navigation, not to try and compensate for a component in need of refactoring.

I agree with @kwight's assessment above.

> All in all, I think we're back with two options: the current one, Rodrigo's full Redux.

If you remember the video I posted in the channel, there might also be a third option of passing shared logic and state to child components via render props. :smile: Or maybe splitting everything into container and presentational (stateful and pure) components if you think that would make it more readable.

**Comment by kwight** [comment]
> Unless... we applied it to a component containing both List and Navigation. Such component currently exists, and it's the List itself.

I took a look and basically did this ^ – but it's the higher-up `CommentsManagement` (which we can just call `Comments`) that acts as the HOC. And you're right, @Copons , it makes _that_ component now the 600LOC file rather than `CommentsList` (which becomes a simple component listing out whatever comments are passed to it). While not magically reducing lines of code, this has the big advantage of making comments the same architecture as posts and pages, and much easier to understand when going to fix something: `Comments` is the HOC managing state and defining actions, while everything below just receive props (and `connect()`s for anything else it needs from the store, like maybe post info for `CommentDetail`? Didn't get that far).

I feel like with this more traditional format and duties in place, further refactoring will be easier to implement and understand – without having to bring in the store with no real justification. This demo PR #19075 shuffles these pieces around to illustrate (it displays comments, but actions are properly wired up).

**Comment by gwwar** [comment]
I'm putting some of the larger refactors on hold for now. Folks also had some concerns on if global state was the best fit for this use case.

**Comment by rodrigoi** [comment]
Closed, since I don't think we'll revisit this, although the conversation was productive 😉
